spectacles/runner.py

```diff
@@ -8,119 +10,161 @@
-        self.import_projects = import_projects
 
-        # Get the current branch so we can return to it afterwards
-        self.original_branch = self.client.get_active_branch_name(self.project)
-        self.temp_branches: List[BranchState] = []
+        state: ProjectState = self.get_project_state()
+        self.workspace: str = state.workspace
+        self.history: List[ProjectState] = [state]
+        self.imports: List[str] = self.get_project_imports()
+        logger.debug(
+            f"Project '{self.project}' imports the following projects: {self.imports}"
+        )
+
+        self.commit: Optional[str] = None
+        self.branch: Optional[str] = None
+        self.is_temp_branch: bool = False
+        self.import_managers: List[LookerBranchManager] = []
+
+    def __call__(
+        self,
+        branch: Optional[str] = None,
+        commit: Optional[str] = None,
+        ephemeral: Optional[bool] = None,
+    ):
+        if branch and commit:
+            raise ValueError("Cannot call with both branch and commit.")
+        self.branch = branch
+        self.commit = commit
+        self.ephemeral = ephemeral or bool(commit)
+        return self
 
     def __enter__(self):
-        self.client.update_workspace(self.project, self.workspace)
-        if self.commit_ref:
-            """Can't delete branches from the production workspace so we need to save
-            the starting dev branch to return to and to use as a base to delete
-            temporary branches from."""
-            starting_branch = self.client.get_active_branch_name(self.project)
-            temp_branch = self.setup_temp_branch(
-                self.project, original_branch=starting_branch
-            )
-            self.client.create_branch(self.project, temp_branch)
-            self.client.update_branch(self.project, temp_branch, self.commit_ref)
-        # If we didn't start on the desired branch, check it out
-        elif self.original_branch != self.name:
-            self.client.checkout_branch(self.project, self.name)
+        # A branch was passed, so we check it out in dev mode.
+        if self.branch:
+            self.update_workspace("dev")
+            self.client.checkout_branch(self.project, self.branch)
             if self.remote_reset:
                 self.client.reset_to_remote(self.project)
-        if self.import_projects and self.name != "master":
-            self.branch_imported_projects()
+        # A commit was passed, so we non-destructively create a temporary branch we can
+        # hard reset to the commit.
+        elif self.commit:
+            self.branch = self.checkout_temp_branch(self.commit)
+        # Neither branch nor commit were passed, so go to production.
+        else:
+            if self.init_state.workspace == "production":
+                prod_state = self.init_state
+            else:
+                self.update_workspace("production")
+                prod_state = self.get_project_state()
```

Comment:
I don't think this matters, but is it actually possible to start an API session _not_ in production mode?

Reply:
No, but we might be doing something like incremental mode where we're re-using the same branch manager and thus the same session.